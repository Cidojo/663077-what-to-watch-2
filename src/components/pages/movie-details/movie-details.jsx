import * as React from 'react';
import * as PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';

import {MovieCardBackground} from './../../parts/movie-card-background/movie-card-background.jsx';
import {Header} from './../../parts/header/header.jsx';
import {Footer} from './../../parts/footer/footer.jsx';
import {MovieCardButtons} from './../../parts/movie-card-buttons/movie-card-buttons.jsx';
import {MovieCardMeta} from './../../parts/movie-card-meta/movie-card-meta.jsx';
import {MovieCardPoster} from './../../parts/movie-card-poster/movie-card-poster.jsx';
import Catalog from './../../parts/catalog/catalog.jsx';
import Player from './../../parts/player/player.jsx';
import MovieCardTabWithActiveTab from './../../parts/movie-card-tabs/movie-card-tabs.jsx';
import withPlayerScreen from './../../../hocs/with-player-screen/with-player-screen.jsx';

import {movieCardPropTypes} from './../../../global-custom-types.js';
import Selectors from './../../../selectors/selectors.js';
import {AuthActionCreator} from './../../../reducers/auth-reducer/auth-reducer.js';

const MovieDetails = (props) => {
  const {
    activeCard,
    relatedMovies,
    genre,
    onShowPlayer,
    onClosePlayer,
    isPlayerShown,
    isAuthorizationRequired,
    requireAuth
  } = props;

  if (!activeCard) {
    return undefined;
  }

  const {
    id,
    name,
    backgroundImage,
    released,
    isFavorite,
    posterImage
  } = activeCard;

  // requireAuth();

  if (isAuthorizationRequired) {
    return (
      <Redirect to="/login"/>
    );
  }

  if (isPlayerShown) {
    return (
      <div className="player">
        <Player
          card={activeCard}
          muted={true}
          autoplay={true}
          controls={true}
          onClosePlayer={onClosePlayer}
        />
      </div>
    );
  }

  return (
    <>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <MovieCardBackground
            name={name}
            backgroundImage={backgroundImage}
          />

          <h1 className="visually-hidden">WTW</h1>

          <Header
            extraClassName="movie-card__head"
            isWithUserBlock={true}
          />

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <MovieCardMeta
                name={name}
                genre={genre}
                released={released}
              />
              <MovieCardButtons
                movieId={id}
                isFavorite={isFavorite}
                onPlayButtonClick={onShowPlayer}
              />
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">

            <MovieCardPoster
              classMods="movie-card__poster--big"
              name={name}
              posterImage={posterImage}
            />

            <MovieCardTabWithActiveTab
              movie={activeCard}
              // comments={comments}
            />
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <Catalog
            movieCards={relatedMovies}
          />
          <Footer/>
        </section>
      </div>
    </>
  );
};

MovieDetails.propTypes = {
  activeCard: movieCardPropTypes,
  genre: PropTypes.string,
  relatedMovies: PropTypes.array,
  isPlayerShown: PropTypes.bool,
  onShowPlayer: PropTypes.func,
  onClosePlayer: PropTypes.func,
  isAuthorizationRequired: PropTypes.bool,
  requireAuth: PropTypes.func
};

MovieDetails.defaultProps = {
  genre: ``,
  activeCard: {},
  relatedMovies: [],
  isPlayerShown: false,
  onShowPlayer: () => {},
  onClosePlayer: () => {},
  isAuthorizationRequired: false,
  requireAuth: () => {}
};

const MovieDetailsWrapped = withPlayerScreen(MovieDetails);

const mapDispatchToProps = (dispatch) => ({
  requireAuth: () => dispatch(AuthActionCreator.requiredAuthorization(true))
});

const mapStateToProps = (state, ownProps) => {
  return Object.assign({}, ownProps, {
    isAuthorizationRequired: Selectors.getAuthRequiredStatus(state),
    relatedMovies: Selectors.getRelatedMovies(state),
  });
};

export {MovieDetails};
export default connect(mapStateToProps, mapDispatchToProps)(MovieDetailsWrapped);
