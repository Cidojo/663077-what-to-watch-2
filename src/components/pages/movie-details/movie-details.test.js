import * as React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {shallow, configure} from 'enzyme';
import toJson from 'enzyme-to-json';
import {MovieDetails} from './movie-details.jsx';

configure({adapter: new Adapter()});

const props = {
  activeCard: {},
  relatedMovies: [],
  isPlayerShown: false,
  onShowPlayer: () => {},
  onClosePlayer: () => {},
  onLoadReviews: () => {},
  reviews: {
    id: 0,
    user: {
      id: 0,
      name: ``,
    },
    rating: 0,
    comment: 0,
    date: ``
  },
  match: {
    params: {
      id: `0`
    }
  }
};

it(`should render MovieDetails component without errors`, () => {
  const tree = shallow(<MovieDetails {...props}/>);
  expect(toJson(tree)).toMatchSnapshot();
});
