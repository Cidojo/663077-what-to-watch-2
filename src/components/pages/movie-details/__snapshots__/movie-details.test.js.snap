// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieDetails component without errors 1`] = `
<Fragment>
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <MovieCardBackground
        backgroundImage=""
        name=""
      />
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <Header
        extraClassName="movie-card__head"
        isWithUserBlock={true}
        title=""
      />
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <MovieCardMeta
            genre=""
            name=""
            released={0}
          />
          <Connect(MovieCardButtons)
            onPlayButtonClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <MovieCardPoster
          classMods="movie-card__poster--big"
          extraClassName=""
          name=""
          posterImage=""
        />
        <WithActiveItem
          movie={Object {}}
          reviews={Array []}
        />
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <Connect(Catalog)
        movieCards={Array []}
      />
      <Footer />
    </section>
  </div>
</Fragment>
`;
